### DISCUSSION TITLE
SimpleTcpClient: Can DataReceived be async?

### DISCUSSION THREAD

**Discussion by woha** [jchristn/SuperSimpleTcp#210] [Q&A]
I want to await an async call inside my `DataReceived` method and want to make `DataReceived` itself async.

In the change notes under "New in v3.0.x", I found "Configurable sync vs async for firing DataReceived events (thank you @TheNybbler)", but I was unable to find further documentation or samples.

Is there a way to await async calls in `DataReceived`?

**Comment by jchristn** [comment]
The `Settings.UseAsyncDataReceivedEvents`, when set to `true`, simply fires the event handler as an unawaited task.  I'm not aware of a widely agreed-upon standard for async event handlers.

**Reply by sherman89** [reply]
@jchristn Thanks for this library!

Forgive me if I'm misunderstanding something, but couldn't DataReceived theoretically be called in the wrong order because the task is not awaited?

```csharp
Action action = () => _events.HandleDataReceived(this, new DataReceivedEventArgs(ServerIpPort, data));
if (_settings.UseAsyncDataReceivedEvents)
{
    _ = Task.Run(action, token);
}
else
{
    action.Invoke();
}
```
